torch/csrc/Exceptions.h

```diff
@@ -349,26 +349,35 @@ struct PyWarningHandler {
 };
 
 namespace detail {
+
+struct noop_gil_scoped_release {
+  // user-defined constructor (i.e. not defaulted) to avoid
+  // unused-variable warnings at usage sites of this class
+  noop_gil_scoped_release() {}
+};
+
+template <bool release_gil>
+using conditional_gil_scoped_release = std::conditional_t<
+    release_gil,
+    pybind11::gil_scoped_release,
+    noop_gil_scoped_release>;
+
 template <typename Func, size_t i>
 using Arg = typename invoke_traits<Func>::template arg<i>::type;
 
-template <typename Func, size_t... Is>
+template <typename Func, size_t... Is, bool release_gil>
 auto wrap_pybind_function_impl_(
     // NOLINTNEXTLINE(cppcoreguidelines-missing-std-forward)
     Func&& f,
     std::index_sequence<Is...>,
-    bool release_gil) {
+    std::bool_constant<release_gil>) {
   namespace py = pybind11;
 
   // f=f is needed to handle function references on older compilers
-  return [f = std::forward<Func>(f), release_gil](Arg<Func, Is>... args) {
+  return [f = std::forward<Func>(f)](Arg<Func, Is>... args) {
     HANDLE_TH_ERRORS
-    if (release_gil) {
-      py::gil_scoped_release no_gil;
-      return c10::guts::invoke(f, std::forward<Arg<Func, Is>>(args)...);
-    } else {
-      return c10::guts::invoke(f, std::forward<Arg<Func, Is>>(args)...);
-    }
+    conditional_gil_scoped_release<release_gil> no_gil;
```

Comment:
Couldn't you avoid this whole conditional_gil_scoped_release class with just an if constexpr?

Reply:
The point of this no-op scope guard was to not have a branch where all the other code would be duplicated